flower/events.py

```diff
@@ -39,31 +37,46 @@ def __init__(self, *args, **kwargs):
         self.metrics = PrometheusMetrics()
 
     def event(self, event):
+        # Save the event
+        super(EventsState, self).event(event)
+
         worker_name = event['hostname']
         event_type = event['type']
 
         self.counter[worker_name][event_type] += 1
 
         if event_type.startswith('task-'):
             task_id = event['uuid']
+            task = self.tasks.get(task_id)
             task_name = event.get('name', '')
             if not task_name and task_id in self.tasks:
-                task_name = self.tasks[task_id].name or ''
+                task_name = task.name or ''
             self.metrics.events.labels(worker_name, event_type, task_name).inc()
 
             runtime = event.get('runtime', 0)
             if runtime:
                 self.metrics.runtime.labels(worker_name, task_name).observe(runtime)
 
+            task_started = task.started
+            task_received = task.received
```

Comment:
I have followed what is done so far in this method and did not split it into sub-methods/sub-handlers as I can see the `Receiver` object takes only 1 handler and I guess for a good reason.